src/lib.rs

```diff
@@ -346,45 +353,97 @@ mod test {
     }
 
     #[test]
-    fn udp_assosiate_no_auth() {
-        env_logger::init();
+    fn mock_udp_assosiate_no_auth() {
+        init();
         block_on(async {
+            const MOCK_ADDRESS: &str = "[::1]:40235";
+
             let (tx, rx) = oneshot::channel();
             tokio::spawn(setup_socks_server("[::1]:0", None, tx));
             let backing_socket = TcpStream::connect(rx.await.unwrap()).await.unwrap();
 
             let tunnel = client::Socks5Datagram::bind(backing_socket, "[::]:0")
                 .await
                 .unwrap();
+            let mock_udp_server = UdpSocket::bind(MOCK_ADDRESS).await.unwrap();
 
-            associate(tunnel, "[::1]:40235")
+            tunnel
+                .send_to(
+                    b"hello world!",
+                    MOCK_ADDRESS.to_socket_addrs().unwrap().next().unwrap(),
+                )
                 .await
-                .expect("error on associate");
+                .unwrap();
+            println!("Send packet to {}", MOCK_ADDRESS);
+
+            let mut buf = [0; 13];
+            let (len, addr) = mock_udp_server.recv_from(&mut buf).await.unwrap();
+            assert_eq!(len, 12);
+            assert_eq!(&buf[..12], b"hello world!");
+
+            mock_udp_server
+                .send_to(b"hello world!", addr)
+                .await
+                .unwrap();
+
+            println!("Recieve packet from {}", MOCK_ADDRESS);
+            let len = tunnel.recv_from(&mut buf).await.unwrap().0;
+            assert_eq!(len, 12);
+            assert_eq!(&buf[..12], b"hello world!");
         });
     }
 
-    async fn associate(socks: client::Socks5Datagram<TcpStream>, mock_server: &str) -> Result<()> {
-        let socket = UdpSocket::bind(mock_server).await?;
-
-        socks
-            .send_to(
-                b"hello world!",
-                mock_server.to_socket_addrs()?.next().unwrap(),
-            )
-            .await?;
-        println!("Send packet to {}", mock_server);
+    #[test]
+    fn dns_udp_assosiate_no_auth() {
+        init();
+        block_on(async {
+            const DNS_SERVER: &str = "1.1.1.1:53";
 
-        let mut buf = [0; 13];
-        let (len, addr) = socket.recv_from(&mut buf).await?;
-        assert_eq!(len, 12);
-        assert_eq!(&buf[..12], b"hello world!");
+            let (tx, rx) = oneshot::channel();
+            tokio::spawn(setup_socks_server("[::1]:0", None, tx));
+            let backing_socket = TcpStream::connect(rx.await.unwrap()).await.unwrap();
 
-        socket.send_to(b"hello world!", addr).await?;
+            let tunnel = client::Socks5Datagram::bind(backing_socket, "[::]:0")
```

Comment:
more comment is required here 🙏

Reply:
Sure!

I've added additional comments here, and also added additional documentation and examples at the Socks5Datagram::bind definition.